hal/spi_api.h

```diff
@@ -19,220 +18,346 @@
 #ifndef MBED_SPI_API_H
 #define MBED_SPI_API_H
 
+#include <stdbool.h>
+#include <stdint.h>
 #include "device.h"
-#include "hal/dma_api.h"
-#include "hal/buffer.h"
 
 #if DEVICE_SPI
+/**
+ * \defgroup hal_spi SPI: Serial peripheral interface HAL API.
+ * Low level interface to the serial peripheral interface of a target.
+ *
+ * ## Programming model
+ * This is an OOP like API. spi_t is used a "this".<br/>
+ * Methods not having this type as their first parameter are **class** methods.<br/>
+ * Methods with an spi_t object as their first parameter are **instance** methods.<br/>
+ * `spi_init()` acts as the "constructor"(/initialiser) of this *class*.<br/>
+ * `spi_free()` acts as the "destructor"(/deinitialiser) of this *class*.<br/>
+ *
+ * ## General requirements
+ * - Multiple instances can share pins such as MISO/MOSI/MCLK but not their SS pin. This pin is used
+ *   to distinguish communication channels (master or slaves).
+ * - As the instance allocation is at the user's responsibility, the hal implementation cannot keep
+ *   a copy of the pointer (keep an alias to the object) as it may be stack allocated and/or moved.
+ * - All method must be thread safe at HAL level.
+ * - The hal implementation must support both clock phases.
+ * - The hal implementation must support both clock polarities.
+ * - The hal implementation must support either master or slave or both modes.
+ * - The hal implementation must support either lsb first or msb first or both bit ordering.
+ * - The hal implementation must support either continuous or noncontinuous or both modes.
+ * - The hal implementation must support at least 1 word length in the range 4 to 32.
+ * - The hal implementation must support the simplex and full duplex modes.
+ * - The hal implementation may or may not support the half-duplex, dual and/or quad-io modes.
+ * - The accuracy of the clock generated in master mode by the hal implementation must be within +/-
+ *   0.1% of the user requested frequency.
+ *
+ * # Defined behaviours
+ * - spi_get_capabilities(..)
```

Comment:
Please reference the test function which will be verifying each of the defined behaviors (or indicate that the behavior is not verified). You'll probably need to add a lot of test prototypes.

An example of this can be found in the RTC API:
https://github.com/ARMmbed/mbed-os/blob/2f37602451d789526201ddf5c6733b94f57e9ad6/hal/rtc_api.h#L38-L47

Reply:
These functions are not yet defined. @mprse and @jamesbeyond are working on this.